samza-test/src/main/java/org/apache/samza/test/framework/stream/InMemoryInputDescriptor.java

```diff
@@ -0,0 +1,126 @@
+    Preconditions.checkNotNull(systemDescriptor);
+    Preconditions.checkNotNull(streamId);
+  }
+
+  /**
+   * Using this method on an InMemoryInputDescriptor creates a single partitioned inmemory stream
+   * and initializes the stream with messages
+   *
+   * @param partitionData messages used to initialize the single partition stream
+   * @return this input descriptor
+   */
+  public InMemoryInputDescriptor<StreamMessageType> withData(
+      List<StreamMessageType> partitionData) {
+    Preconditions.checkNotNull(partitionData);
+    Map<Integer, Iterable<StreamMessageType>> partition = new HashMap<>();
+    partition.put(0, partitionData);
+    initializeInMemoryInputStream(partition);
+    return this;
+  }
+
+  /**
+   * Using this method on an InMemoryInputDescriptor creates a multi-partitioned inmemory stream
+   * and initializes partitions with stream with messages
+   *
+   * @param partitionData key of the map represents partitionId and value represents
+   *                      messages in the partition
+   * @return this input descriptor
+   */
+  public InMemoryInputDescriptor<StreamMessageType> withData(
+      Map<Integer, ? extends Iterable<StreamMessageType>> partitionData) {
+    Preconditions.checkNotNull(partitionData);
+    initializeInMemoryInputStream(partitionData);
+    return this;
+  }
+
+  @Override
+  public Map<String, String> toConfig() {
+    HashMap<String, String> configs = new HashMap<>(super.toConfig());
+    InMemorySystemDescriptor descriptor = (InMemorySystemDescriptor) getSystemDescriptor();
+    configs.put(InMemorySystemConfig.INMEMORY_SCOPE, descriptor.getInMemoryScope());
+    return configs;
+  }
+
+  /**
+   * Creates an in memory stream with {@link InMemorySystemFactory} and initializes the metadata for the stream.
+   * Initializes each partition of that stream with messages
+   * @param partitions represents the descriptor describing a stream to initialize with the in memory system
+   */
+  private void initializeInMemoryInputStream(Map<Integer, ? extends Iterable<StreamMessageType>> partitions) {
+    String systemName = getSystemName();
+    String physicalName = (String) getPhysicalName().orElse(getStreamId());
+    StreamSpec spec = new StreamSpec(getStreamId(), physicalName, systemName, partitions.size());
+    SystemFactory factory = new InMemorySystemFactory();
+    factory.getAdmin(systemName, new MapConfig(toConfig())).createStream(spec);
+    SystemProducer producer = factory.getProducer(systemName, new MapConfig(toConfig()), null);
+    SystemStream sysStream = new SystemStream(systemName, physicalName);
+    partitions.forEach((partitionId, partition) -> {
+        partition.forEach(e -> {
+            Object key = e instanceof KV ? ((KV) e).getKey() : null;
+            Object value = e instanceof KV ? ((KV) e).getValue() : e;
+            producer.send(systemName, new OutgoingMessageEnvelope(sysStream, Integer.valueOf(partitionId), key, value));
```

Comment:
Just to confirm: Does this work because the in-memory static map allows you to produce input here and then consume with the in-memory consumer (from the in-memory system factory config override) in the job?